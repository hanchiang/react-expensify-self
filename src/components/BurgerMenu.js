import React from 'react';
import { slide as Menu } from 'react-burger-menu';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

export class BurgerMenu extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false
    };
  }

  isMenuOpen(state) {
    return state.isOpen;
  }

  render() {
    return (
      <Menu right isOpen={this.state.isOpen} onStateChange={this.isMenuOpen}>
          <Link to="/profile">Profile</Link>
          <button className="button button-link logout-button" type="button" onClick={this.props.startLogout}>Logout</button>
      </Menu>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(null, mapDispatchToProps)(BurgerMenu);