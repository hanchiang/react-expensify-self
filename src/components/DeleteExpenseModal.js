import React from 'react';
import Modal from 'react-modal';

Modal.setAppElement('#app');

function DeleteExpenseModal(props) {
  const handleConfirmDelete = (event) => props.onModalConfirm(true);
  const handleRejectDelete = (event) => props.onModalConfirm(false);

  return (
    <Modal
      isOpen={props.showModal}
      onRequestClose={props.onModalClose}
      closeTimeoutMS={300}
      contentLabel="Delete expense?"
      className="modal"
    >
      <div className="modal-header">
        <h2 className="modal-title">Delete Expense</h2>
      </div>

      <div className="modal-body">
        <p className="modal-message">Are you sure you want to delete expense?</p>
      </div>

      <div className="modal-actions">
        <button className="button button-normal" onClick={handleRejectDelete} type="button">Close</button>
        <button className="button button-warning" onClick={handleConfirmDelete} type="button">Delete</button>  
      </div>
    </Modal>
  );

}

export default DeleteExpenseModal;