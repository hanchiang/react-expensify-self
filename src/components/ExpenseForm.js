import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    
    const { expense } = this.props;
    this.state = {
      error: '',
      description: expense ? expense.description : '',
      amount: expense ? expense.amount/100 : '',
      note: expense ? expense.note : '',
      createdAt: expense ? moment(expense.createdAt) : moment(),
      focused: false
    };
  }

  onDateChange = (createdAt) => this.setState({createdAt});
  onFocusChange = ({focused}) => this.setState({focused});
  onDescriptionChange = (event) => this.setState({description: event.target.value});
  onAmountChange = (event) => {
    const amount = event.target.value;
    const amountRegex = /^\d+(\.\d{0,2})?$/;
    if (!amount || amount.match(amountRegex)) {
      this.setState({amount});
    }
  }
  onNoteChange = (event) => this.setState({ note: event.target.value });

  onSubmit = (event) => {
    event.preventDefault();
    let error = '';

    if (!this.state.description || !this.state.amount) {
      error = 'Please enter description and amount';
      this.setState({error});
    } else {
      this.setState({error});
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  }

  render() {
    const { error } = this.state;
    return (
        <form onSubmit={this.onSubmit} className="form">
          {error && <div className="form-error">{error}</div>}

          <input className="text-input" type="text" value={this.state.description} onChange={this.onDescriptionChange} placeholder="description" autoFocus />
          <input className="text-input" type="text" value={this.state.amount} onChange={this.onAmountChange} placeholder="Amount" />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea className="textarea" value={this.state.note} onChange={this.onNoteChange} placeholder="Add a note for your expense (optional)"></textarea>

          <div>
            <button className="button">Add Expense</button>
          </div>
        </form>
    );
  }
}

export default ExpenseForm;