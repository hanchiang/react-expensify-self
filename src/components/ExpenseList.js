import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/getVisibleExpenses';

function ExpenseList(props) {
  const { expenses } = props;

  return (
    <div>
      <div className="list-header">
        <div>Expenses</div>
        <div className="show-for-desktop">Amount</div>
      </div>

      <div className="list-body">
        {
          expenses.length === 0 ? <div className="list-item list-item-message">No expenses</div>
          :
          expenses.map(expense =>
            <ExpenseListItem key={expense.id} expense={expense} />
          )
        }
      </div>
    </div>
  )
}

const mapStateToProps = (state) => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);