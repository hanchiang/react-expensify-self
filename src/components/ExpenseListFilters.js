import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      focusedInput: null
    };
  }
  handleTextFilterChange = (event) => this.props.setTextFilter(event.target.value);
  handleSortFilterChange = (event) => this.props.setSortBy(event.target.value);
  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = (focusedInput) => this.setState({focusedInput});


  render () {
    return (
      <div>
        <div className="input-group">
          <div className="input-group-item">
            <input type="text" className="text-input" onChange={this.handleTextFilterChange} placeholder="Search expense" />
          </div>

          <div className="input-group-item">
            <select value={this.props.sortBy} className="select" onChange={this.handleSortFilterChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>

          <div className="input-group-item date-range-picker">
            <DateRangePicker
              startDate={this.props.startDate}
              endDate={this.props.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusedInput}
              onFocusChange={this.onFocusChange}
              startDateId={'1'}
              endDateId={'2'}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}


const mapStateToProps = (state) => ({
  sortBy: state.filters.sortBy,
  startDate: state.filters.startDate,
  endDate: state.filters.endDate
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (value) => dispatch(setTextFilter(value)),
  setSortBy: (value) => value === 'date' ? dispatch(sortByDate()) : dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);