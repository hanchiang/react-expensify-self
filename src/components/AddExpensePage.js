import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import moment from 'moment';

export function AddExpensePage(props) {
  const onSubmit = (expense) => {
    props.onSubmit(expense);
    props.history.push('/dashboard');
  }

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header-title">Add Expense</h1>
        </div>
      </div>
      
      <div className="content-container">
        <ExpenseForm onSubmit={onSubmit} />
      </div>
      
    </div>
  );
}

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (expense) => dispatch(startAddExpense(expense))
});

export default connect(null, mapDispatchToProps)(AddExpensePage);