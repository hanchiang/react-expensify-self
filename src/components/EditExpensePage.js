import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import DeleteExpenseModal from './DeleteExpenseModal';

class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showModal: false,
      shouldRemoveExpense: false
    };
    this.onSubmit = this.onSubmit.bind(this);
    this.onRemoveClick = this.onRemoveClick.bind(this);
    this.onModalClose = this.onModalClose.bind(this);
    this.onModalConfirm = this.onModalConfirm.bind(this);
  }
  onSubmit(expense) {
    console.log('submit!')
    this.props.onSubmit(expense);
    this.props.history.push('/dashboard');
  }

  onRemoveClick() {
    this.setState({ showModal: true });
  }

  onModalClose() {
    this.setState({ showModal: false });
  }

  onModalConfirm(shouldRemoveExpense) {
    this.setState({ shouldRemoveExpense, showModal: false });
    if (shouldRemoveExpense) {
      this.props.onRemove(this.props.expense.id);
      this.props.history.push('/dashboard');
    }
  }

  render() {
    const { expense } = this.props;
    return (
      <div id="edit-expense-page">
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header-title">Edit Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm expense={expense} onSubmit={this.onSubmit} isEdit={true} onRemove={this.onRemoveClick} />
        </div>

      {
        <DeleteExpenseModal showModal={this.state.showModal}
          onModalConfirm={this.onModalConfirm}
          onModalClose={this.onModalClose}
        />
      }
      </div>
    );
  }
  
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  onSubmit: (expense) => dispatch(startEditExpense(props.match.params.id, expense)),
  onRemove: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);