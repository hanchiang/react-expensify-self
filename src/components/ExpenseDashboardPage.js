import React from 'react';
import { NavLink } from 'react-router-dom';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

function ExpenseDashboardPage() {
    return (
        <div>
            <h1>Expense dashboard page</h1>

            <div className="page-header">
                <div className="content-container">
                    <ExpensesSummary />
                    <NavLink className="button" to="/create">Add expense</NavLink>
                </div>
            </div>

            <ExpenseListFilters />
            <ExpenseList />
            
        </div>
    )
}

export default ExpenseDashboardPage;