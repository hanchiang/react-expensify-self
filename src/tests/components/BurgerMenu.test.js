import React from 'react';
import { shallow } from 'enzyme';

import { BurgerMenu } from '../../components/BurgerMenu';

describe('BurgerMenu', () => {
  it('should render BurgerMenu', () => {
    const wrapper = shallow(<BurgerMenu />);
    expect(wrapper).toMatchSnapshot();
  });
})