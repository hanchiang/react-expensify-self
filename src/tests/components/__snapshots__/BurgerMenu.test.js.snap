// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BurgerMenu should render BurgerMenu 1`] = `
<Menu
  bodyClassName=""
  burgerBarClassName=""
  burgerButtonClassName=""
  className=""
  crossButtonClassName=""
  crossClassName=""
  id=""
  isOpen={false}
  itemListClassName=""
  menuClassName=""
  morphShapeClassName=""
  noOverlay={false}
  onStateChange={[Function]}
  outerContainerId=""
  overlayClassName=""
  pageWrapId=""
  right={true}
  styles={Object {}}
  width={300}
>
  <Link
    replace={false}
    to="/profile"
  >
    Profile
  </Link>
  <button
    className="button button-link logout-button"
  >
    Logout
  </button>
</Menu>
`;
